### PR TITLE
[action-gateway] Add pouring action

### PR DESCRIPTION
**Pull request by jchevrie**
Add a pouring action callable via rpc `pour ("target" x y z) ("neck_position" x y z)`:

- The target coordinates `("target" x y z)` are the top of the receiving container (glass) expressed in the reference frame.
- The neck_offset coordinates `("neck_position" x y z)` are the position of the neck of the container (bottle) expressed in the grasper frame (hand frame).
The pouring motion will perform a rotation around this point.

Parameters that can be changed via the config file:
- `H_offset` is an horizontal offset between the target and the bottle neck
- `V_offset` is a vertical offset between the target and the bottle neck
- `init_inclin` is the initial inclination from the vertical (in degrees) of the bottle at the hand of the approach phase, before performing the pouring rotation.
- `final_inclin` is the final inclination from the vertical at the end of the pouring motion.

This function can only work after the `grasp` function has already been called, and it will move the corresponding hand.

The code is working in simulation but still needs testing on the robot.

### CODE REVIEW

File: modules/action-gateway/src/main.cpp [hunk 1]
```diff
@@ -754,6 +762,127 @@ class Gateway : public RFModule
...
+    bool pouringApproach(const Vector &target, const Vector &neck_position,
+                         const string &part="select")
+    {
+        if ((target.length()<3) || (neck_position.length()!=3))
+        {
+            yError()<<"Invalid parameters given for pouring approach";
+            return false;
+        }
+
+        std:string part_=choosePart(part, target);
```
Comment (pattacini) [inline modules/action-gateway/src/main.cpp]: this should be `std::string`, although I would rather use `string` plainly to be consistent with the rest of the code here.

File: modules/action-gateway/src/main.cpp [hunk 2]
```diff
@@ -754,6 +762,127 @@ class Gateway : public RFModule
...
+
+        Vector hand_pose(7);
+        hand_pose.setSubvector(0,hand_position);
+        hand_pose.setSubvector(3, hand_orientation);
+
+        pouring.current_hand_pose = hand_pose;
+        return reach(pouring.current_hand_pose,part_);
+    }
+
+    bool pouringMotion(const Vector &neck_position, double angle, const string &part="select")
```
Comment (pattacini) [inline modules/action-gateway/src/main.cpp]: missing `/****************************************************************/`

File: modules/action-gateway/src/main.cpp [hunk 3]
```diff
@@ -992,6 +1121,15 @@ class Gateway : public RFModule
             grasping.lift=gGrasping.check("lift",Value(0.1)).asDouble();
         }
 
+        Bottle &gPouring=rf.findGroup("pouring");
+        if (!gPouring.isNull())
+        {
+            pouring.H_offset=gPouring.check("H_offset",Value(0.01)).asDouble();
+            pouring.V_offset=gPouring.check("",Value(0.02)).asDouble();
```
Comment (pattacini) [inline modules/action-gateway/src/main.cpp]: It should `check("V_offset",...))`

File: modules/action-gateway/app/conf/config.ini [hunk 4]
```diff
@@ -21,6 +21,12 @@ left_hand           (70.0 70.0)     // [deg]
 right_hand          (70.0 70.0)     // [deg]
 lift                0.15            // [m]
 
+[pouring]
```
Comment (pattacini) [inline modules/action-gateway/app/conf/config.ini]: `config-sim.ini` needs to be updated too.

**Comment by jchevrie** [comment]
Thanks for the review of the typos, code formatting and source of bugs.

Regarding the signature, I find `"destination"` less accurate than `"target"` since the bottle will not really end up at this destination, but the robot will try to pour the water to the target (adding some offset to the actual bottle position).
But it can make sense to change it.

For the change of `"neck_position"` into `"source"`, however, it seems less clear to me, because it is not the starting position of the bottle in the reference frame (which, to me, would be implied by the wording `source` and `destination`).
"neck_position"` is a property that only depends on the shape of the container.
Or maybe you see it as the `source` of the water, i.e. the container, and in this case it would make sense.

Anyway, these are only placeholders and I don't check them in the code (should I?), so there is no problem changing them.
Let me know what is the best.

**Comment by pattacini** [comment]
Yep, with `source` and `destination` I readily meant the source and destination (maybe **sink** is even better?) of the liquid itself 😉

**Comment by pattacini** [comment]
You could conveniently avoid checking the placeholders in the code as done now 👍

**Comment by jchevrie** [comment]
Ok. I will simply change the order of the parameters then (since `destination` is currently the first parameter and `source` the last one).

**Pull request merged**